### ISSUE TITLE
Cannot import Keras model: mis matched lengths / output is not an IOutputLayer / model cannot be trained

### ISSUE THREAD

**Issue by iRazvan** [deeplearning4j/deeplearning4j#4005]
Hi, I'm having a lot of issues importing a simple keras model.
The model is here: https://gist.github.com/iRazvan/3cd24dc5b6e5f49e0d928b3b7ad2d237

Keras version: 1.1.0
Tensorflow version: 0.12.1
DL4J version: 0.9.1

The code for DL4J is actually based on the examples:

```
String modelHDF5Path = "/media/razvan/FC202B84202B454C/Razvan/CNN/VP/new_model3.hdf5";
MultiLayerNetwork nn = KerasModelImport.importKerasSequentialModelAndWeights(modelHDF5Path);
        log.info(nn.summary());
```

I get this error:
```
org.deeplearning4j.examples.multigpu.TestImportKeras
o.n.l.f.Nd4jBackend - Loaded [JCublasBackend] backend
o.n.n.NativeOpsHolder - Number of threads used for NativeOps: 32
o.n.l.a.o.e.DefaultOpExecutioner - Backend used: [CUDA]; OS: [Linux]
o.n.l.a.o.e.DefaultOpExecutioner - Cores: [8]; Memory: [14.0GB];
o.n.l.a.o.e.DefaultOpExecutioner - Blas vendor: [CUBLAS]
o.n.l.j.o.e.CudaExecutioner - Device name: [GeForce GTX 1080 Ti]; CC: [6.1]; Total/free memory: [11711807488]
o.n.l.j.o.e.CudaExecutioner - Device name: [GeForce GTX 1080 Ti]; CC: [6.1]; Total/free memory: [11715084288]
o.d.n.m.k.KerasSequentialModel - Model cannot be trained: output layer dense_3_loss is not an IOutputLayer (no loss function specified)
o.d.n.c.l.LayerValidation - Layer "batchnormalization_1" momentum has been set but will not be applied unless the updater is set to NESTEROVS.
o.d.n.m.MultiLayerNetwork - Starting MultiLayerNetwork with WorkspaceModes set to [training: NONE; inference: SEPARATE]
Exception in thread "main" java.lang.IllegalStateException: Mis matched lengths: [3] != [48]
	at org.nd4j.linalg.util.LinAlgExceptions.assertSameLength(LinAlgExceptions.java:40)
	at org.nd4j.linalg.api.ops.BaseTransformOp.<init>(BaseTransformOp.java:47)
	at org.nd4j.linalg.api.ops.impl.transforms.Set.<init>(Set.java:24)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.assign(BaseNDArray.java:1267)
	at org.deeplearning4j.nn.layers.BaseLayer.setParam(BaseLayer.java:218)
	at org.deeplearning4j.nn.modelimport.keras.KerasLayer.copyWeightsToLayer(KerasLayer.java:475)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.helperCopyWeightsToModel(KerasModel.java:701)
	at org.deeplearning4j.nn.modelimport.keras.KerasSequentialModel.getMultiLayerNetwork(KerasSequentialModel.java:227)
	at org.deeplearning4j.nn.modelimport.keras.KerasSequentialModel.getMultiLayerNetwork(KerasSequentialModel.java:213)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasSequentialModelAndWeights(KerasModelImport.java:157)
	at org.deeplearning4j.examples.multigpu.TestImportKeras.main(TestImportKeras.java:69)

Process finished with exit code 1

```

Any ideas? I'm stuck with this for a while now.. found nothing useful in the forums. My end goal would be to import the model in Android.

**Comment by maxpumperla** [comment]
@iRazvan I'll look into this soon and will include your case in integration tests for reference.

In the mean time, would you mind stepping through the debugger and telling me at which layer, for which weight the assignment of weights fails?

**Comment by iRazvan** [comment]
Crashes at batchnormalization_1, I have attached a screenshot. 
I have also attached the saved more here: https://drive.google.com/drive/folders/0B_Hdsrv9K7NDTjBVbkNFOE14WVU?usp=sharing

I have tried numerous ways to import it, still no luck.

```
MultiLayerNetwork graph;
graph = KerasModelImport.importKerasSequentialModelAndWeights(modelJsonPath, weightsH5Path, false);
```
       

The code above gives the same error message.

**Comment by maxpumperla** [comment]
hey @iRazvan thanks for all the details, very helpful. The problem is this: keras has an `axis` argument that we currently do not support. You want to batch along `axis=1`, which indeed has length 48. We currently only support batching of the _last_ batch dimension (here length 3).

I'd have to go back and extend our batchnorm layer to really address this, which might take a while. Not sure if I have time for this right now.

In the meantime, what you can do is this: reshape your input so that your batch-dimension is in fact last. Might have to shuffle dims again afterwards.

**Comment by maxpumperla** [comment]
@iRazvan at least on master we now have an informative error message for your use case.

@AlexDBlack what's your take here? Would it take a lot of effort to support batch normalisation along other axes than the last?

**Comment by AlexDBlack** [comment]
batch norm in DL4J operates along dimension 1 (depth for CNNs, or values per  time step for time series, or each column for dense).
default ordering of data in keras CNNs is channels_last - i.e., [N, H, W, C]. Is there a use case where batch norm along the height dimension (sharing across width/channels) makes sense? If that's the layout (it's not explicitly configured) then it seems invalid to me...
https://keras.io/layers/convolutional/
that said, if it is channels_first [N, C, H, W] (which is also what DL4J uses) then dimension 1 makes sense and it should work out of the box...

As for supporting batch norm along other axes: I'v *very* hestitant. Show me a published paper that actually does something other than what we do, and I'd consider it :)

**Comment by maxpumperla** [comment]
@AlexDBlack cool, that's fine with me. I also don't really see a real-world use case here. will make sure to double check that we take into account the two data formats properly in model import, though.

**Comment by iRazvan** [comment]
My default ordering was: [batch size, height, width, channels], I changed it in keras batch norm and at conv2d to use dim_ordering = 'th' and now it's compatible with dl4j [batch size, channels, height, width]. 
Thanks for the help!
